app/code/Magento/SalesRule/Model/Quote/Discount.php

```diff
@@ -88,19 +56,16 @@ class Discount extends AbstractTotal
      * @param RulesApplier|null $rulesApplier
      */
     public function __construct(
-        ManagerInterface $eventManager,
-        StoreManagerInterface $storeManager,
+        protected readonly ManagerInterface $eventManager,
+        protected readonly StoreManagerInterface $storeManager,
         Validator $validator,
-        PriceCurrencyInterface $priceCurrency,
-        RuleDiscountInterfaceFactory $discountInterfaceFactory = null,
-        DiscountDataInterfaceFactory $discountDataInterfaceFactory = null,
-        RulesApplier $rulesApplier = null
+        protected readonly PriceCurrencyInterface $priceCurrency,
+        private ?RuleDiscountInterfaceFactory $discountInterfaceFactory = null,
```

Comment:
The type of this argument has changed by mistake.
```suggestion
        private RuleDiscountInterfaceFactory $discountInterfaceFactory = null,
```